File: 104-maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {

        function<int(TreeNode*, int)>dfs = [&](TreeNode* node, int count){
            if (!node) return count;

            count += 1;
            int left = dfs(node->left,count);
            int right = dfs(node->right,count);

            return max(left,right);
        };
        return dfs(root,0);
    

        
    }
};

File: 104-maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def maxDepth(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        def dfs(node,count):
            if node:
                count +=1
            if not node:
                return count
            
            left = dfs(node.left,count)
            right = dfs(node.right,count)

            return max(left,right)
        
        return dfs(root,0)
        